src/cunumeric/matrix/diag.cc

```diff
@@ -22,30 +22,55 @@ namespace cunumeric {
 using namespace Legion;
 using namespace legate;
 
-template <LegateTypeCode CODE>
-struct DiagImplBody<VariantKind::CPU, CODE> {
+template <LegateTypeCode CODE, int DIM>
+struct DiagImplBody<VariantKind::CPU, CODE, DIM, true> {
   using VAL = legate_type_of<CODE>;
 
-  void operator()(const AccessorWO<VAL, 2>& out,
-                  const AccessorRO<VAL, 2>& in,
-                  const Point<2>& start,
-                  size_t distance) const
+  void operator()(const AccessorRD<SumReduction<VAL>, true, DIM>& out,
+                  const AccessorRO<VAL, DIM>& in,
+                  const coord_t& start,
+                  Pitches<DIM - 1>& m_pitches,
+                  Rect<DIM>& m_shape,
+                  const size_t naxes,
+                  const coord_t distance) const
   {
-    for (coord_t idx = 0; idx < distance; ++idx) {
-      Point<2> p(start[0] + idx, start[1] + idx);
-      out[p] = in[p];
+    size_t skip_size = 1;
+
+    for (int i = 0; i < naxes; i++) {
+      auto diff = 1 + m_shape.hi[DIM - i - 1] - m_shape.lo[DIM - i - 1];
+      if (diff != 0) skip_size *= diff;
+    }
+    const size_t volume = m_shape.volume();
+    for (size_t idx = 0; idx < volume; idx += skip_size) {
+      Point<DIM> in_p  = m_pitches.unflatten(idx, m_shape.lo);
+      Point<DIM> out_p = in_p;
+      for (coord_t d = 0; d < distance; ++d) {
+        for (size_t i = DIM - naxes; i < DIM; i++) {
+          in_p[i]  = start + d;
+          out_p[i] = start + d;
+        }
+        auto v = in[in_p];
+        out.reduce(out_p, v);
+      }
     }
   }
+};
+
+// not extract (create a new 2D matrix with diagonal from vector)
+template <LegateTypeCode CODE>
+struct DiagImplBody<VariantKind::CPU, CODE, 2, false> {
+  using VAL = legate_type_of<CODE>;
 
-  void operator()(const AccessorRD<SumReduction<VAL>, true, 2>& out,
-                  const AccessorRO<VAL, 2>& in,
-                  const Point<2>& start,
-                  size_t distance) const
+  void operator()(const AccessorRO<VAL, 2>& in,
+                  const AccessorRW<VAL, 2>& out,
+                  const int offset,
+                  const Point<2> start,
+                  const coord_t distance)
   {
-    for (coord_t idx = 0; idx < distance; ++idx) {
-      Point<2> p(start[0] + idx, start[1] + idx);
-      auto v = in[p];
-      out.reduce(p, v);
+    for (coord_t idx = 0; idx < distance; idx++) {
+      Point<2> in_p(start[0] + idx, start[1] + idx);
+      Point<2> out_p(start[0] + idx, start[1] + idx);
```

Comment:
It looks like these two points are the same? Is there a reason not to use the same variable for both? This seems to be the case in the `extract == True` case as well.

Reply:
No reason, just a leftover from previous implementation